cpp/ql/test/library-tests/syntax-zoo/dataflow-ir-consistency.expected

```diff
@@ -1464,113 +1468,1235 @@ unreachableNodeCCtx
+| test.c:204:12:204:20 | VariableAddress [post update] | PostUpdateNode should not be the target of local flow. |
+| test.c:220:5:220:13 | VariableAddress [post update] | PostUpdateNode should not be the target of local flow. |
+| test.c:222:3:222:11 | VariableAddress [post update] | PostUpdateNode should not be the target of local flow. |
+| test.c:227:5:227:13 | VariableAddress [post update] | PostUpdateNode should not be the target of local flow. |
+| test.c:229:3:229:11 | VariableAddress [post update] | PostUpdateNode should not be the target of local flow. |
+| test.c:233:7:233:15 | VariableAddress [post update] | PostUpdateNode should not be the target of local flow. |
+| test.c:233:7:233:15 | VariableAddress [post update] | PostUpdateNode should not be the target of local flow. |
+| try_catch.cpp:7:8:7:8 | Argument this [post update] | PostUpdateNode should not be the target of local flow. |
+| try_catch.cpp:7:8:7:8 | Argument this [post update] | PostUpdateNode should not be the target of local flow. |
+| try_catch.cpp:7:8:7:8 | ConvertToNonVirtualBase [post update] | PostUpdateNode should not be the target of local flow. |
+| try_catch.cpp:7:8:7:8 | ConvertToNonVirtualBase [post update] | PostUpdateNode should not be the target of local flow. |
+| try_catch.cpp:7:8:7:8 | this [post update] | PostUpdateNode should not be the target of local flow. |
+| try_catch.cpp:7:8:7:8 | this [post update] | PostUpdateNode should not be the target of local flow. |
+| try_catch.cpp:13:5:13:16 | Argument this [post update] | PostUpdateNode should not be the target of local flow. |
+| try_catch.cpp:13:5:13:16 | VariableAddress [post update] | PostUpdateNode should not be the target of local flow. |
+| try_catch.cpp:21:14:21:20 | VariableAddress [post update] | PostUpdateNode should not be the target of local flow. |
+| try_catch.cpp:23:8:23:14 | VariableAddress [post update] | PostUpdateNode should not be the target of local flow. |
+| unaryopexpr.c:10:5:10:5 | VariableAddress [post update] | PostUpdateNode should not be the target of local flow. |
+| unaryopexpr.c:11:5:11:5 | VariableAddress [post update] | PostUpdateNode should not be the target of local flow. |
+| unaryopexpr.c:12:7:12:7 | VariableAddress [post update] | PostUpdateNode should not be the target of local flow. |
+| unaryopexpr.c:13:7:13:7 | VariableAddress [post update] | PostUpdateNode should not be the target of local flow. |
+| void_bug.cpp:5:3:5:11 | VariableAddress [post update] | PostUpdateNode should not be the target of local flow. |
+| whilestmt.c:9:7:9:10 | VariableAddress [post update] | PostUpdateNode should not be the target of local flow. |
+| whilestmt.c:11:5:11:8 | done [post update] | PostUpdateNode should not be the target of local flow. |
+| whilestmt.c:40:7:40:7 | VariableAddress [post update] | PostUpdateNode should not be the target of local flow. |
+| whilestmt.c:42:7:42:7 | VariableAddress [post update] | PostUpdateNode should not be the target of local flow. |
```

Comment:
Yes. All these consistency issues come from the `StoreNodeFlow::flowThrough` predicate which does `PostUpdateNode -> PostUpdateNode` flow to handle all the different kinds of conversions happening between fields lookups. The intention of `PostUpdateNode`s is that they should only arise as the target of a `storeStep`, and should not be produced by `simpleLocalFlowStep`. However, in order to skip these (from the purpose of dataflow) irrelevant dataflow nodes that represent conversions, we need to step over them.

We could get rid of these at the cost of adding another "data flow in data flow" mechanism, but as @jbj alluded to [https://github.com/github/codeql/pull/6825#pullrequestreview-796179710](here) this is probably not a solution we want.

As far as I know, no one is aware of any performance issues caused by having `PostUpdate -> PostUpdate` simple flow steps. The consistency check was just put in since such steps aren't present in the other language's dataflow implementation.

Reply:
Perhaps we should remove that particular consistency check (for C++)?

Reply:
The consistency queries are shared via `identical-files` between all the languages. I can think of a number of ways to handle this:
- Opt-out of `identical-files` for this file
- Add a pyrameterized predicate to modify the behavior of this check on a per-language basis

I'm much more in favor of the second option. We have an instance of this approach already with the  [`isImmutableOrUnobservable`](https://github.com/github/codeql/blob/main/cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll#L497) predicate. I'm not exactly sure about the interface for this pyrameterized predicate would look like, though.

Reply:
Before we jump through such hoops, I think you should talk to the other data flow maintainers about whether flow into post-update nodes can sometimes be a good thing. [Last time I checked](https://github.com/github/codeql/pull/3389#issue-610759281), Java could not send post-update flow through casts because it was missing flow into post-update nodes.

Reply:
Indeed. It might be that this consistency check should be relaxed _for all languages_.